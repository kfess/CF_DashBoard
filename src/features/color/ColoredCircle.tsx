import React from "react";
import { styled } from "@mui/system";

const StyledCircle = styled("span", {
  shouldForwardProp: (prop) => prop !== "color" && prop !== "fillPercent",
})<{
  color: string;
  fillPercent?: number;
  width?: string;
  height?: string;
}>(({ theme, color, fillPercent = 1, width, height }) => ({
  borderStyle: "solid",
  borderWidth: "thin",
  display: "inline-block",
  borderRadius: "50%",
  padding: "5px",
  margin: "0 5px",
  width: width,
  height: height,
  borderColor: color,
  background: `linear-gradient(to top, ${color} ${
    (fillPercent ?? 1) * 100
  }%, rgba(0,0,0,0) ${(fillPercent ?? 1) * 100}%)`,
}));

type Props = {
  readonly color: string;
  readonly fillPercent?: number;
  readonly width?: string;
  readonly height?: string;
};

export const ColoredCircle: React.FC<Props> = (props: Props) => {
  const { color, fillPercent, width, height } = props;

  return (
    <StyledCircle
      aria-label="colored circle"
      color={color}
      fillPercent={fillPercent}
      width={width}
      height={height}
    />
  );
};
