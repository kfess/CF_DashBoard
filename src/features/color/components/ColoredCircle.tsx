import React from "react";
import { styled } from "@mui/system";

const StyledCircle = styled("span", {
  shouldForwardProp: (prop) => prop !== "color" && prop !== "fillPercent",
})<{
  color: string;
  fillPercent?: number;
  width?: string;
  height?: string;
}>(({ color, fillPercent = 1 }) => ({
  borderStyle: "solid",
  borderWidth: "thin",
  display: "inline-block",
  borderRadius: "50%",
  padding: "5px",
  margin: "0 5px",
  borderColor: color,
  background: `linear-gradient(to top, ${color} ${
    (fillPercent ?? 1) * 100
  }%, rgba(0,0,0,0) ${(fillPercent ?? 1) * 100}%)`,
}));

type Props = {
  readonly color: string;
  readonly fillPercent?: number;
};

export const ColoredCircle: React.FC<Props> = (props: Props) => {
  const { color, fillPercent } = props;

  return (
    <StyledCircle
      aria-label="colored circle"
      color={color}
      fillPercent={fillPercent}
    />
  );
};
