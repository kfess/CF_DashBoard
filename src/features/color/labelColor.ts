declare const hexaColorNominality: unique symbol;
export type HexaColor = string & { [hexaColorNominality]: never };

export const isValidHexaColor = (value: string): boolean =>
  /^#[0-9a-fA-F]{6}$/.test(value) || /^#[0-9a-fA-F]{3}$/.test(value);

export const generateRandomHexaColor = (): HexaColor => {
  const highContrastColor = (): string => {
    let value = Math.floor(Math.random() * 256);
    return value < 220 ? value.toString(16).padStart(2, "0") : "00";
  };
  return `#${highContrastColor()}${highContrastColor()}${highContrastColor()}` as HexaColor;
};

// HSL to HEX 変換関数
const hslToHex = (h: number, s: number, l: number) => {
  let r, g, b;
  if (s === 0) {
    r = g = b = l;
  } else {
    const hue2rgb = (p: number, q: number, t: number) => {
      if (t < 0) t += 1;
      if (t > 1) t -= 1;
      if (t < 1 / 6) return p + (q - p) * 6 * t;
      if (t < 1 / 2) return q;
      if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
      return p;
    };
    const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
    const p = 2 * l - q;
    r = hue2rgb(p, q, h + 1 / 3);
    g = hue2rgb(p, q, h);
    b = hue2rgb(p, q, h - 1 / 3);
  }
  const toHex = (x: number) => {
    const hex = Math.round(x * 255).toString(16);
    return hex.length === 1 ? "0" + hex : hex;
  };
  return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
};

// 高コントラストかつカラフルな色を生成する関数
export const generateHighContrastColor = () => {
  const h = Math.random();
  const s = 0.8 + Math.random() * 0.2;
  const l = 0.4 + Math.random() * 0.2;
  return hslToHex(h, s, l);
};
