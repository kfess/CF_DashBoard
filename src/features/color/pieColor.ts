export const pieColors = [
  "#4169e1",
  "#FF6D6A",
  "#48C774",
  "#FFD54F",
  "#FF9800",
  "#9C27B0",
  "#26C6DA",
  "#FF4081",
  "#9E9D24",
  "#FFB2B2",
  "#00695C",
  "#E1BEE7",
  "#8D6E63",
  "#B71C1C",
  "#B2DFDB",
  "#FDD835",
  "#3F51B5",
  "#616161",
  "#000000",
  "#2E7D32",
  "#FF867C",
  "#64B5F6",
  "#FFAB40",
  "#7E57C2",
  "#D4E157",
  "#FF8A80",
  "#26A69A",
  "#5C6BC0",
  "#9E9E9E",
  "#D32F2F",
  "#81C784",
  "#7E57C2",
  "#FFEE58",
  "#5E35B1",
  "#FFA726",
  "#8D6E63",
  "#2E7D32",
  "#5C6BC0",
  "#FFAB00",
  "#5E35B1",
];
