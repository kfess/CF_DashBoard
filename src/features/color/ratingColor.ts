import { ThemeColor } from "@features/color/@types/theme";

export const ratingColor = [
  "Black",
  "Gray",
  "Green",
  "Cyan",
  "Blue",
  "Violet",
  "LightOrange",
  "DeepOrange",
  "LightRed",
  "Red",
  "DeepRed",
] as const;
export type RatingColor = typeof ratingColor[number];

// for theme color (base, purple, green)
type ColorCode =
  | "#000000" // Black
  | "#CCCCCC" // Gray
  | "#76DEBB" // Green
  | "#76FF77" // Cyan
  | "#AAAAFF" // Blue
  | "#FF88FF" // Violet
  | "#FFCC87" // LightOrange
  | "#FFBB55" // DeepOrange
  | "#FF7777" // LightRed
  | "#FF3333" // Red
  | "#AA0100"; // DeepRed

// for therme color (dark)
type DarkColorCode =
  | "#202020" // Black
  | "#CCCCCC" // Gray // need Change!
  | "#76DEBB" // Green // need Change!
  | "#76FF77" // Cyan // need Change!
  | "#AAAAFF" // Blue // need Change!
  | "#FF88FF" // Violet // need Change!
  | "#FFCC87" // LightOrange // need Change!
  | "#FFBB55" // DeepOrange // need Change!
  | "#FF7777" // LightRed // need Change!
  | "#FF3333" // Red // need Change!
  | "#AA0100"; // DeepRed // need Change!

type Title =
  | "Legendary Grandmaster" // DeepRed
  | "International Grandmaster" // Red
  | "Grandmaster" // LightRed
  | "International Master" // DeepOrange
  | "Master" // LightOrange
  | "Candidate Master" // Violet
  | "Expert" // Blue
  | "Specialist" // Cyan
  | "Pupil" // Green
  | "Newbie" // Gray
  | "No Data"; // Black

export type RatingColorInfo = {
  [C in RatingColor]: {
    readonly name: C;
    readonly colorCode: ColorCode;
    readonly darkColorCode: DarkColorCode;
    readonly lowerBound: number;
    readonly upperBound: number;
    readonly title: Title;
  };
};

const MAX_RATING = 5000; // max rating of Codeforces?
const NO_RATING = -1;

const ratingColorInfo: RatingColorInfo = {
  Black: {
    name: "Black",
    colorCode: "#000000",
    darkColorCode: "#202020",
    lowerBound: NO_RATING,
    upperBound: NO_RATING,
    title: "No Data",
  },
  Gray: {
    name: "Gray",
    colorCode: "#CCCCCC",
    darkColorCode: "#CCCCCC",
    lowerBound: 0,
    upperBound: 1199,
    title: "Newbie",
  },
  Green: {
    name: "Green",
    colorCode: "#76DEBB",
    darkColorCode: "#76DEBB",
    lowerBound: 1200,
    upperBound: 1399,
    title: "Pupil",
  },
  Cyan: {
    name: "Cyan",
    colorCode: "#76FF77",
    darkColorCode: "#76FF77",
    lowerBound: 1400,
    upperBound: 1599,
    title: "Specialist",
  },
  Blue: {
    name: "Blue",
    colorCode: "#AAAAFF",
    darkColorCode: "#AAAAFF",
    lowerBound: 1600,
    upperBound: 1899,
    title: "Expert",
  },
  Violet: {
    name: "Violet",
    colorCode: "#FF88FF",
    darkColorCode: "#FF88FF",
    lowerBound: 1900,
    upperBound: 2099,
    title: "Candidate Master",
  },
  LightOrange: {
    name: "LightOrange",
    colorCode: "#FFCC87",
    darkColorCode: "#FFCC87",
    lowerBound: 2100,
    upperBound: 2299,
    title: "Master",
  },
  DeepOrange: {
    name: "DeepOrange",
    colorCode: "#FFBB55",
    darkColorCode: "#FFBB55",
    lowerBound: 2300,
    upperBound: 2399,
    title: "International Master",
  },
  LightRed: {
    name: "LightRed",
    colorCode: "#FF7777",
    darkColorCode: "#FF7777",
    lowerBound: 2400,
    upperBound: 2599,
    title: "Grandmaster",
  },
  Red: {
    name: "Red",
    colorCode: "#FF3333",
    darkColorCode: "#FF3333",
    lowerBound: 2600,
    upperBound: 2999,
    title: "International Grandmaster",
  },
  DeepRed: {
    name: "DeepRed",
    colorCode: "#AA0100",
    darkColorCode: "#AA0100",
    lowerBound: 3000,
    upperBound: MAX_RATING,
    title: "Legendary Grandmaster",
  },
};

export const getRatingColorInfo = (rating: number | undefined) => {
  if (rating == null) {
    return ratingColorInfo["Black"];
  } else if (rating <= ratingColorInfo["Gray"].upperBound) {
    return ratingColorInfo["Gray"];
  } else if (rating <= ratingColorInfo["Green"].upperBound) {
    return ratingColorInfo["Green"];
  } else if (rating <= ratingColorInfo["Cyan"].upperBound) {
    return ratingColorInfo["Cyan"];
  } else if (rating <= ratingColorInfo["Blue"].upperBound) {
    return ratingColorInfo["Blue"];
  } else if (rating <= ratingColorInfo["Violet"].upperBound) {
    return ratingColorInfo["Violet"];
  } else if (rating <= ratingColorInfo["LightOrange"].upperBound) {
    return ratingColorInfo["LightOrange"];
  } else if (rating <= ratingColorInfo["DeepOrange"].upperBound) {
    return ratingColorInfo["DeepOrange"];
  } else if (rating <= ratingColorInfo["LightRed"].upperBound) {
    return ratingColorInfo["LightRed"];
  } else if (rating <= ratingColorInfo["Red"].upperBound) {
    return ratingColorInfo["Red"];
  } else {
    return ratingColorInfo["DeepRed"];
  }
};

const isDarkThemeColor = (themeColor: ThemeColor) => themeColor === "dark";

export const getColorCodeFromRating = (
  rating: number | undefined,
  themeColor: ThemeColor = "base" // default theme
): ColorCode | DarkColorCode => {
  if (rating == null) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Black"].colorCode
      : ratingColorInfo["Black"].darkColorCode;
  } else if (rating <= ratingColorInfo["Gray"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Gray"].colorCode
      : ratingColorInfo["Gray"].darkColorCode;
  } else if (rating <= ratingColorInfo["Green"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Green"].colorCode
      : ratingColorInfo["Green"].darkColorCode;
  } else if (rating <= ratingColorInfo["Cyan"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Cyan"].colorCode
      : ratingColorInfo["Cyan"].darkColorCode;
  } else if (rating <= ratingColorInfo["Blue"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Blue"].colorCode
      : ratingColorInfo["Blue"].darkColorCode;
  } else if (rating <= ratingColorInfo["Violet"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Violet"].colorCode
      : ratingColorInfo["Violet"].darkColorCode;
  } else if (rating <= ratingColorInfo["LightOrange"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["LightOrange"].colorCode
      : ratingColorInfo["LightOrange"].darkColorCode;
  } else if (rating <= ratingColorInfo["DeepOrange"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["DeepOrange"].colorCode
      : ratingColorInfo["DeepOrange"].darkColorCode;
  } else if (rating <= ratingColorInfo["LightRed"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["LightRed"].colorCode
      : ratingColorInfo["LightRed"].darkColorCode;
  } else if (rating <= ratingColorInfo["Red"].upperBound) {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["Red"].colorCode
      : ratingColorInfo["Red"].darkColorCode;
  } else {
    return !isDarkThemeColor(themeColor)
      ? ratingColorInfo["DeepRed"].colorCode
      : ratingColorInfo["DeepRed"].darkColorCode;
  }
};

// To Do
export const calcFillPercent = (rating: number | undefined) => {
  if (rating == null) {
    return 1;
  } else if (rating <= ratingColorInfo["Gray"].upperBound) {
    return (rating - 800) / 400;
  } else if (rating < ratingColorInfo["Green"].upperBound) {
    return 1 - (1400 - rating) / 200;
  } else if (rating < ratingColorInfo["Cyan"].upperBound) {
    return 1 - (1600 - rating) / 200;
  } else if (rating < ratingColorInfo["Blue"].upperBound) {
    return 1 - (1900 - rating) / 300;
  } else if (rating < ratingColorInfo["Violet"].upperBound) {
    return 1 - (2100 - rating) / 200;
  } else if (rating < ratingColorInfo["LightOrange"].upperBound) {
    return 1 - (2300 - rating) / 300;
  } else if (rating < ratingColorInfo["DeepOrange"].upperBound) {
    return 1 - (2400 - rating) / 300;
  } else if (rating < ratingColorInfo["Red"].upperBound) {
    return 1 - (2600 - rating) / 200;
  } else {
    return 1;
  }
};
