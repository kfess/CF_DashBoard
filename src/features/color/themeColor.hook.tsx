// import {
//   useState,
//   useContext,
//   createContext,
//   ReactNode,
//   useCallback,
// } from "react";
// import { Theme, ThemeProvider } from "@mui/material/styles";
// import { useTheme } from "@mui/material/styles";
// // import { baseTheme, darkTheme } from "@features/color/themeColor";

// const Context = createContext<{
//   setBaseTheme: () => void;
//   setDarkTheme: () => void;
// }>({
//   setBaseTheme: () => {},
//   setDarkTheme: () => {},
// });

// export const ThemeContextProvider = ({ children }: { children: ReactNode }) => {
//   const [theme, setTheme] = useState<Theme>(baseTheme);

//   const setBaseTheme = useCallback(() => setTheme(baseTheme), [theme]);
//   const setDarkTheme = useCallback(() => setTheme(darkTheme), [theme]);

//   return (
//     <ThemeProvider theme={theme}>
//       <Context.Provider value={{ setBaseTheme, setDarkTheme }}>
//         {children}
//       </Context.Provider>
//     </ThemeProvider>
//   );
// };

// export const useThemeContext = () => {
//   const theme = useTheme();
//   const { setBaseTheme, setDarkTheme } = useContext(Context);

//   return { theme, setBaseTheme, setDarkTheme };
// };
