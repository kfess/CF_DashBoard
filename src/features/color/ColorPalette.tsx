import React from "react";
import { HexaColor } from "@features/color/labelColor";
import { ColoredCircle } from "@features/color/ColoredCircle";

const vividColors = [
  "#FF0000",
  "#FFA500",
  "#FFD700",
  "#008000",
  "#008B8B",
  "#4169E1",
  "#0000FF",
  "#8A2BE2",
] as const;

const paleColors = [
  "#FFB6C1",
  "#FFDEAD",
  "#F0E68C",
  "#9ACD32",
  "#3CB371",
  "#6495ED",
  "#4169E1",
  "#9370DB",
] as const;

type Props = {
  setColor: (arg: HexaColor) => void;
};

export const ColorPalette: React.FC<Props> = (props: Props) => {
  const { setColor } = props;

  return (
    <>
      <div css={{ marginBottom: "0.3rem" }}>Choose from popular colors</div>
      {vividColors.map((c) => {
        return (
          <span
            key={c}
            css={{
              cursor: "pointer",
              marginRight: "0.1rem",
              marginLeft: "0.1rem",
            }}
          >
            <span
              onClick={() => {
                setColor(c as HexaColor);
              }}
            >
              <ColoredCircle color={c} width="0.5rem" height="0.5rem" />
            </span>
          </span>
        );
      })}
      <div></div>
      {paleColors.map((c) => {
        return (
          <span
            key={c}
            css={{
              cursor: "pointer",
              marginRight: "0.1rem",
              marginLeft: "0.1rem",
            }}
          >
            <span
              onClick={() => {
                setColor(c as HexaColor);
              }}
            >
              <ColoredCircle color={c} width="0.5rem" height="0.5rem" />
            </span>
          </span>
        );
      })}
    </>
  );
};
