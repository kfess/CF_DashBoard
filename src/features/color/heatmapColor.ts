import { getColorCodeFromRating } from "@features/color/ratingColor";

const noActivityColor = (isDarkMode: boolean) =>
  isDarkMode ? "#2d333b" : "#ebedf0";

export const heatmapColors = [
  "#c6e48b", // 最も薄い緑
  "#7bc96f", // 中間の緑
  "#239a3b", // 暗い緑
  "#196127", // 最も濃い緑
] as const;

type ValueToColorMapping = {
  maxVal?: number;
  color: string;
};

const valueToColorMappings: ValueToColorMapping[] = [
  { maxVal: 1, color: heatmapColors[0] }, // Light green
  { maxVal: 3, color: heatmapColors[1] }, // Medium green
  { maxVal: 5, color: heatmapColors[2] }, // Dark green
  { color: heatmapColors[3] }, // Very dark green
];

export const valueToColor = (
  value?: number,
  isDarkMode: boolean = false
): string => {
  if (value === undefined) return noActivityColor(isDarkMode);
  for (const mapping of valueToColorMappings) {
    if (!mapping.maxVal || value < mapping.maxVal) {
      return mapping.color;
    }
  }
  return heatmapColors[3];
};

export const maxDifficultyToColor = (
  value?: number,
  isDarkMode = false
): string => {
  return value === undefined
    ? noActivityColor(isDarkMode)
    : getColorCodeFromRating(value);
};
