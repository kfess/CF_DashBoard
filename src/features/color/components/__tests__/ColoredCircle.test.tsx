import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { ColoredCircle } from "@features/color/ColoredCircle";

describe("ColoredCircle", () => {
  test("renders the circle with the specified color", () => {
    render(<ColoredCircle color="#FF0000" />);
    const circleElement = screen.getByLabelText("colored circle");
    expect(circleElement).toHaveStyle({
      borderColor: "#FF0000",
    });
  });
});
