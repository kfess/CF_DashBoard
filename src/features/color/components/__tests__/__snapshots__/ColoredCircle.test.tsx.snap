// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColoredCircle matches the snapshot with the specified fillPercent 1`] = `
<DocumentFragment>
  <span
    aria-label="colored circle"
    class="css-66aem9"
  />
</DocumentFragment>
`;
